dali/util/cufile_helper.h

```diff
@@ -28,35 +29,60 @@
 // dali device guard
 #include "dali/core/dynlink_cufile.h"
 #include "dali/core/device_guard.h"
+#include "dali/core/cuda_error.h"
 
-// we need this class to make sure that the driver
-// is only opened once per thread. It is not thread safe, coordination outside
 namespace cufile {
 
-// wrapper struct to conveniently store the fd's as well
-class DLL_PUBLIC CUFileHandle{
- public:
-  CUFileHandle() {
-    fd = -1;
-    fdd = -1;
+struct CUFileDriverScope {
+  CUFileDriverScope() {
+    // v2 API performs proper reference counting, so we increase the reference count here...
+    if (cuFileIsSymbolAvailable("cuFileDriverClose_v2")) {
+      CUDA_CALL(cuFileDriverOpen());
+    }
+  }
+  ~CUFileDriverScope() {
+    // ...and decrease it here.
+    // The old GDS API would simply destroy the library, possibly still in use by other modules
+    // within the process.
+    if (cuFileIsSymbolAvailable("cuFileDriverClose_v2")) {
+      CUDA_DTOR_CALL(cuFileDriverClose());  // termination on exception is expected
+    }
   }
+};
 
+// wrapper struct to conveniently store the fd's as well
+class DLL_PUBLIC CUFileHandle {
+ public:
   ~CUFileHandle() {
     Close();
   }
+  CUFileHandle() = default;
+  CUFileHandle(CUFileHandle &&other) {
+    *this = std::move(other);
+  }
+
+  CUFileHandle &operator=(CUFileHandle &&other) {
+    std::swap(fd, other.fd);
+    std::swap(fdd, other.fdd);
+    std::swap(cufh, other.cufh);
+    other.Close();
```

Comment:
Do you have to explicitly call `Close`? Wouldn't it just be called when destructing `other`

Reply:
I read somewhere that implementing move assignment as a pure swap isn't a good practice, because it prolongs the life of the object being overwritten in a surprising way. This is especially important in case of resources like files, which have global side effects. Consider:
```C++
File a = File::open("a.dat", "rw");
File b = File::open("b.dat", "rw");

a = std::move(b);  // the user would expect that "a.dat" is closed, but this assignment actually did a swap
File c = File::open("a.dat", "rw");  // file busy, because the moved-out object `b` is keeping the file open.
```

Reply:
Huh. That's interesting. I didn't know how the move works, then. I thought `b` would be destroyed just after evaluation of this assignment.

Reply:
We learn something new every day ;) The object still "lives", even though it shouldn't be used for anything other than destruction. I'm not sure if the compiler is allowed to destroy a moved-out local variable before the scope end, but there are certainly cases where you can move-out objects with more permanent storage - so they cannot be destroyed (calling a destructor multiple times is an error). For example, you can loop over an array and move every element from it to some other container and then destroy the array. The destructors will be called when the array is destroyed, not when each object in the array is moved out.